padrino-core/lib/padrino-core/path_router/compiler.rb

```diff
@@ -34,12 +34,23 @@ def compiled?
         !!@compiled
       end
   
+
+      ##
+      # Finds routes by using request or env.
+      #
+      def find_by(request_or_env)
+        request = request_or_env.is_a?(Hash) ? Sinatra::Request.new(request_or_env) : request_or_env
+        pattern = encode_default_external(request.path_info)
+        verb    = request.request_method.downcase.to_sym
+        rotation { |offset| match?(offset, pattern) }.select { |route| route.verb == verb }
+      end
+
       ##
-      # Finds routes by using request.
+      # Calls routes by using request.
       #
-      def find_by_request(request)
+      def call_by_request(request)
         rotation do |offset|
-          pattern  = request.path_info.encode(Encoding.default_external)
+          pattern  = encode_default_external(request.path_info)
           if route = match?(offset, pattern)
             params = route.params_for(pattern, request.params)
             yield(route, params) if route.verb == request.request_method.downcase.to_sym
```

Comment:
After this, `route` keeps on being returned regardless it matching the verb or not. Maybe I'm missing something but if we only run its block when the verb matches, why do we need to return it anyway?